### PR TITLE
Merkle tree upgrades

### PR DESCRIPTION
**Pull request by wintersteiger**
Some minor upgrades to the Merkle tree API.

### CODE REVIEW

**Comment by wintersteiger** [comment]
Last of the Merkle tree memory leaks fixed, depending on https://github.com/FStarLang/FStar/pull/2142.

**Comment by msprotz** [comment]
the diff seems a bit big and has a lot of whitespace; any chance you could remove anything in dist that's not merkletree from this PR? thanks!

**Comment by wintersteiger** [comment]
That's what I originally tried to do but some of the tests didn't pass CI, so I checked in a full snapshot. You can set the file filter on the "Files Changed" tab to hide the .c/.h/.ml files though.

**Comment by wintersteiger** [comment]
Hm. The CI produces verification failures that I didn't get locally, so I'll have to start over and I can put the snapshot into a separate commit.

**Comment by msprotz** [comment]
The snapshot is auto-refreshed every night so you can also elect to leave it out... your call! :)

**Comment by wintersteiger** [comment]
That would mean that I have to un-fix `tests/merkle_tree_test.c` because it's compiled against an old `dist`... I'll check again whether it passes CI with just the Merkle tree files in the snapshots.

**Comment by wintersteiger** [comment]
Found another potential problem: `LowStar.RVector.free_elems` is recursive and exhaust the stack space on my dev VM (8MB) while running the hacl-star benchmarks. I may have to get rid of that recursion.

**Comment by msprotz** [comment]
Thanks. Let's try to merge this first, so that the nightly build of HACL* doesn't break (now that your change has been merged in F*).

**Comment by wintersteiger** [comment]
Yeah, just waiting for the CI to finish. Diff is also smaller now and the tests pass on my machine, so I'm hopeful that the CI will pass too.

**Comment by wintersteiger** [comment]
Darn, CI failed because of the stack exhaustion in the benchmarks. I'll see if I can throttle the benchmarks back for now.

**Comment by wintersteiger** [comment]
We can merge this once the CI is done, but I'm afraid I'll have to fix that recursion soon. It looks like it can't handle 128k nodes in one tree.

File: secure_api/merkle_tree/MerkleTree.Low.Datastructures.fst
Comment (msprotz) [inline secure_api/merkle_tree/MerkleTree.Low.Datastructures.fst:289]: is there an implicit argument that's not erased here, i.e. a hidden #hsz ?

**Review by msprotz — APPROVED** [review]
Looks good to me, thanks for taking care of this. I just had one minor question. Can you ask someone else for a review, e.g. Marina?

File: secure_api/merkle_tree/MerkleTree.Low.Datastructures.fst
Comment (wintersteiger) [reply secure_api/merkle_tree/MerkleTree.Low.Datastructures.fst:289]: I don't think so. It just complained about state and ghost not being applicable at the same time, so I copied this trick from your previous changes. `hsz` is not erased (that's the new state argument in the regional), but `#hsz'` is correctly erased.

**Comment by wintersteiger** [comment]
Sure! Many thanks in advance @polubelova! :-)

**Review by polubelova — APPROVED** [review]

**Pull request merged**
